### PR TITLE
[SPARK-7654][SQL] Move insertInto into reader/writer interface.

### PR DESCRIPTION
**Pull request by rxin**
Also removed partitionBy in DataFrameReader, since specifying it explicitly doesn't provide much benefit over not specifying it.

### CODE REVIEW

**Comment by liancheng** [comment]
LGTM

**Comment by rxin** [comment]
Jenkins, retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 1]
```diff
@@ -151,19 +153,52 @@ final class DataFrameWriter private[sql](df: DataFrame) {
...
-        source,
-        temporary = false,
-        partitioningColumns.map(_.toArray).getOrElse(Array.empty[String]),
-        mode,
-        extraOptions.toMap,
-        df.logicalPlan)
-    df.sqlContext.executePlan(cmd).toRdd
+    if (df.sqlContext.catalog.tableExists(tableName :: Nil)) {
+      if (partitioningColumns.isDefined) {
+        sys.error("Inserting data into an existing partitioned table is not yet supported.")
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Inserting into (dynamically) partitioned table is actually supported now. However, a corner case we don't handle well for now is inserting into a Hive metastore table with static partitions. We don't load newly created partitions back into Hive metastore.

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: Well, double checked, we don't support converting partitioned (Parquet) Hive metastore tables.

**Comment by yhuai** [comment]
Seems https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32930/testReport/junit/org.apache.spark.sql.sources/SimpleTextHadoopFsRelationSuite/saveAsTable___load_____non_partitioned_table___Overwrite/ and https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32930/testReport/junit/org.apache.spark.sql.sources/ParquetHadoopFsRelationSuite/saveAsTable___load_____non_partitioned_table___Overwrite/ are caused by a bug. I will take ta look.

**Comment by rxin** [comment]
Closing in favor of https://github.com/apache/spark/pull/6366

**Pull request closed without merging**